### PR TITLE
#159053990 Add Feature For User Following 

### PR DESCRIPTION
**Pull request by RuthNjeri**
**What does this PR do?**

Add a user following functionality.

**Description of Task to be completed?**

_Scenario 1:_
Only authenticated users should be able to follow and or unfollow other logged in users.

_Scenario 2:_
I need a view to manage those I follow and be able to see those who follow me.
Write the tests for user following including the edge cases.

**How to test?**

1. Clone the project from the branch ft-user-following-159053990
2. https://github.com/andela/ah-scorpion.git
3. Create a virtual environment called envvirtualenv env -p python3
4. Activate the virtual environment source env/bin/activate
5. Install the dependencies pip install -r requirements.txt
6. Register multiple users.
7. Login a user who will follow and unfollow other users
8. Test the follow user endpoint using **PUT** request on Postman `http://127.0.0.1:8000/api/v1/profiles/<username>/follow`
9. Test the unfollow user endpoint using **DELETE** request on Postman `http://127.0.0.1:8000/api/v1/profiles/<username>/follow`
10. Check the logged in user's followers
`http://127.0.0.1:8000/api/v1/profiles/followers/`
11. Check the logged in user's following
`http://127.0.0.1:8000/api/v1/profiles/following/`
12. Run the tests coverage run --source='.' manage.py test

What are the relevant pivotal tracker stories?

[#159053990](https://www.pivotaltracker.com/story/show/159053990)

### CODE REVIEW

**Review by hoslack — APPROVED** [review]

File: authors/apps/profiles/tests/test_following.py [hunk 1]
```diff
@@ -0,0 +1,213 @@
...
+    def test_user_following(self):
+        """
+        Test that a user can follow multiple users.
+        """
+        # Pass token of current user
+        self.client.credentials(HTTP_AUTHORIZATION=self.register_login_current_user())
+        # Follow the first user.
+        follow_url = reverse(
+        'profiles:follow_user', args=['sam'])
+        response = self.client.put(follow_url,format='json')
```
Comment (lennykioko) [inline authors/apps/profiles/tests/test_following.py]: PEP8: you need a space after the comma.

File: authors/apps/profiles/urls.py [hunk 2]
```diff
@@ -1,11 +1,24 @@
 from django.urls import path
 
 from authors.apps.profiles.views import ProfileRetrieveAPIView
+from authors.apps.profiles.views import FollowUserRetrieveAPIView
```
Comment (lennykioko) [inline authors/apps/profiles/urls.py]: PEP8: I think it would have been better to use one import statement to import all four classes

File: authors/apps/profiles/urls.py [hunk 3]
```diff
@@ -1,11 +1,24 @@
...
+from authors.apps.profiles.views import UserFollowingListAPIView
+
 
 app_name = "profiles"
 
 urlpatterns = [
     # url to view the profile of a user, does not require permissions
     path('<str:username>', ProfileRetrieveAPIView.as_view(),
          name='view_profile'),
+    # Url to follow a user
```
Comment (lennykioko) [inline authors/apps/profiles/urls.py]: I think the comments are not necessary given that the names of the endpoints are self-explanatory and obvious.

File: authors/apps/profiles/views.py [hunk 4]
```diff
@@ -1,9 +1,15 @@
 # Create your views here.
 from rest_framework.generics import RetrieveAPIView
+from rest_framework.generics import ListAPIView
```
Comment (lennykioko) [inline authors/apps/profiles/views.py:3]: You could use one import statement here as well on the following imports where there is some repetition.

File: authors/apps/profiles/urls.py (hunk 2 above)
Comment (RuthNjeri) [reply authors/apps/profiles/urls.py]: Thanks Lenny, according to [PEP8](https://www.python.org/dev/peps/pep-0008/) the imports should be separated without a comma.

```
Imports should usually be on separate lines:

Yes: import os
     import sys

No:  import sys, os
```

**Review by davenjeru — COMMENTED** [review]
How about viewing another user's followers/following?
You could allow on endpoints like:
`api/v1/profiles/<username>/followers`
`api/v1/profiles/<username>/following`

File: authors/apps/profiles/tests/test_following.py (hunk 1 above)
Comment (RuthNjeri) [reply authors/apps/profiles/tests/test_following.py]: This is noted, thank you.

**Review by munyili-samuel — APPROVED** [review]
I have tested the PR and it is working as per the ReadMe specifications.
Good Job
